src/dmd/root/serializer.d

```diff
@@ -0,0 +1,217 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (C) 1999-2019 by The D Language Foundation, All Rights Reserved
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/serializer.d, _serializer.d)
+ * Documentation:  https://dlang.org/phobos/dmd_serializer.html
+ * Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/serializer.d
+ */
+module dmd.root.serializer;
+
+import core.internal.traits : Unqual;
+import core.stdc.stdarg;
+
+import dmd.root.outbuffer : OutBuffer;
+import dmd.root.traits;
+
+/**
+ * Serializes the given value to a textual representation.
+ *
+ * Params:
+ *  value = the value to serialize
+ *
+ * Returns: the serialized data
+ */
+string serialize(T)(auto ref T value)
+{
+    OutBuffer buffer;
+    Serializer(&buffer).serialize(value);
+
+    return cast(string) buffer.extractSlice;
+}
+
+///
+struct Serializer
+{
+    private
+    {
+        enum Type
+        {
+            none,
+            basic,
+        }
+
+        enum indentation = 2;
+        int level = 0;
+        bool shouldIdent;
+        bool isTopLevel = true;
+        Type previousType = Type.none;
+        OutBuffer* buffer;
+    }
+
+    /// Initializes the serializer with the given buffer.
+    this(OutBuffer* buffer)
+    {
+        this.buffer = buffer;
+        append("---");
+    }
+
+    /**
+     * Serializes the given value to a textual representation.
+     *
+     * This function should be used when customizing the serialization of a type.
+     *
+     * Params:
+     *  value = the value to serialize
+     */
+    void serialize(T)(ref T value)
+    {
+        const isTopLevel = this.isTopLevel;
+        this.isTopLevel = false;
+
+        static if (isBasicType!T)
+            serializeBasicType(value, isTopLevel);
+        else
+            static assert(false, "Serializing a value of type `" ~ T.stringof ~
+                "` is not supported");
+    }
+
+private:
+
+    /**
+     * Serializes a basic type to a textual representation.
+     *
+     * Params:
+     *  value = the value to serialize
+     *  isTopLevel = indicates if this is the first value to be serialized
+     */
+    void serializeBasicType(T)(T value, bool isTopLevel)
```

Comment:
dmd already has `TypeBasic.toChars()`

Reply:
`TypeBasic` is not used here. Native D types are used.